src/maidsafe/vault/data_manager/data_manager.h

```diff
@@ -0,0 +1,156 @@
+
+#ifndef MAIDSAFE_VAULT_DATA_MANAGER_H_
+#define MAIDSAFE_VAULT_DATA_MANAGER_H_
+
+#include "maidsafe/common/types.h"
+
+#include "maidsafe/vault/utils.h"
+#include "maidsafe/vault/data_manager/database.h"
+
+namespace maidsafe {
+
+namespace vault {
+
+template <typename FacadeType>
+class DataManager {
+ public:
+  DataManager(const boost::filesystem::path& vault_root_dir);
+
+  template <typename DataType>
+  routing::HandleGetReturn HandleGet(const routing::SourceAddress& from, const Identity& name);
+
+  template <typename DataType>
+  routing::HandlePutPostReturn HandlePut(const routing::SourceAddress& from,
+                                         const DataType& data);
+
+  template <typename DataType>
+  routing::HandlePutPostReturn
+  HandlePutResponse(const typename DataType::Name& name, const routing::DestinationAddress& from,
+                    const maidsafe_error& return_code);
+
+  void HandleChurn(const routing::CloseGroupDifference& difference);
+
+ private:
+  template <typename DataType>
+  routing::HandlePutPostReturn Replicate(const typename DataType::Name& name,
+                                         const routing::DestinationAddress& exclude);
+
+  void DownRank(const routing::DestinationAddress& /*address*/) {}
+
+  DataManagerDatabase db_;
+  routing::CloseGroupDifference close_group_;
+};
+
+template <typename FacadeType>
+DataManager<FacadeType>::DataManager(const boost::filesystem::path& vault_root_dir)
+    : db_(UniqueDbPath(vault_root_dir)) {}
+
+template <typename FacadeType>
+template <typename DataType>
+routing::HandlePutPostReturn DataManager<FacadeType>::HandlePut(
+    const routing::SourceAddress& /*from*/, const DataType& data) {
+  if (!db_.Exist<DataType>(data.name())) {
+     auto pmid_addresses(static_cast<FacadeType*>(this)
+                             ->template GetClosestNodes<DataType>(data.name()));
+    db_.Put<DataType>(data.name(), pmid_addresses);
+    std::vector<routing::DestinationAddress> dest_addresses;
+    for (const auto& pmid_address : pmid_addresses)
+      dest_addresses.emplace_back(std::make_pair(routing::Destination(pmid_address),
+                                                 boost::none));
+    return dest_addresses;
+  }
+  return boost::make_unexpected(MakeError(CommonErrors::success));
+}
+
+template <typename FacadeType>
+template <typename DataType>
+routing::HandlePutPostReturn DataManager<FacadeType>::HandlePutResponse(
+    const typename DataType::Name& name, const routing::DestinationAddress& from,
+    const maidsafe_error& return_code) {
+  assert(return_code.code() != make_error_code(CommonErrors::success));
+  DownRank(from);  // failed to store
+  return Replicate<DataType>(name, from);
+}
+
+template <typename FacadeType>
+template <typename DataType>
+routing::HandlePutPostReturn
+DataManager<FacadeType>::Replicate(const typename DataType::Name& name,
+                                   const routing::DestinationAddress& from) {
```

Comment:
May be we should have an explicit name to say it is replicating along with removing a pmid name from the db.
ReplicateOnFailure ?